Exception when return type is Iterable

Hi, I have a **Quarkus** application that has **smallrye-open-api** as a dependency.

The application has a resource that contains an endpoint with `Iterable` as return type.
```
@Path("/api/books")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class BookResource {

    @Inject
    BookService bookService;

    @GET
    public Iterable<Book> getBooks() {
        return bookService.getBooks();
    }

    ...
}
```

It seems that **smallrye-open-api** doesn't handle `Iterable` response since because, when I start the application, I get the error below.

> **P.S**. If I replace the return from `Iterable` to `List` or `Set`, everything works fine.

Thanks!

```
19:14:39,858 ERROR [io.qua.dev.DevModeMain] Failed to start Quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.NullPointerException
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.createResponseFromJaxRsMethod(OpenApiAnnotationScanner.java:934)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsMethod(OpenApiAnnotationScanner.java:688)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processJaxRsResourceClass$0(OpenApiAnnotationScanner.java:422)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1548)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:420)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:221)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:298)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:246)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:941)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1535)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1426)
        at java.lang.Thread.run(Thread.java:745)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)

        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:157)
        at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:177)
        at io.quarkus.dev.DevModeMain.start(DevModeMain.java:95)
        at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.NullPointerException
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.createResponseFromJaxRsMethod(OpenApiAnnotationScanner.java:934)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsMethod(OpenApiAnnotationScanner.java:688)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processJaxRsResourceClass$0(OpenApiAnnotationScanner.java:422)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1548)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:420)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:221)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:298)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:246)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:941)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1535)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1426)
        at java.lang.Thread.run(Thread.java:745)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)

        at io.quarkus.builder.Execution.run(Execution.java:108)
        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:121)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:130)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:111)
        ... 3 more
Caused by: java.lang.NullPointerException
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.createResponseFromJaxRsMethod(OpenApiAnnotationScanner.java:934)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsMethod(OpenApiAnnotationScanner.java:688)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processJaxRsResourceClass$0(OpenApiAnnotationScanner.java:422)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1548)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:420)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:221)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:298)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:246)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:941)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1535)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1426)
        at java.lang.Thread.run(Thread.java:745)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

--- a/implementation/src/main/java/io/smallrye/openapi/api/util/ArchiveUtil.java
+++ b/implementation/src/main/java/io/smallrye/openapi/api/util/ArchiveUtil.java
@@ -128,6 +128,7 @@ public class ArchiveUtil {
 
         Indexer indexer = new Indexer();
         index(indexer, "io/smallrye/openapi/runtime/scanner/CollectionStandin.class");
+        index(indexer, "io/smallrye/openapi/runtime/scanner/IterableStandin.class");
         index(indexer, "io/smallrye/openapi/runtime/scanner/MapStandin.class");
         indexArchive(config, indexer, archive);
         return indexer.complete();

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/IterableStandin.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/IterableStandin.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright 2019 Red Hat, Inc, and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.smallrye.openapi.runtime.scanner;
+
+/**
+ * @author Eric Wittmann
+ */
+public abstract class IterableStandin<E> implements Iterable<E> {
+    E value;
+}

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java
@@ -82,6 +82,9 @@ public class OpenApiDataObjectScanner {
     // Collection (list-type things)
     public static final DotName COLLECTION_INTERFACE_NAME = DotName.createSimple(Collection.class.getName());
     public static final Type COLLECTION_TYPE = Type.create(COLLECTION_INTERFACE_NAME, Type.Kind.CLASS);
+    // Iterable (also list-type things)
+    public static final DotName ITERABLE_INTERFACE_NAME = DotName.createSimple(Iterable.class.getName());
+    public static final Type ITERABLE_TYPE = Type.create(ITERABLE_INTERFACE_NAME, Type.Kind.CLASS);
     // Map
     public static final DotName MAP_INTERFACE_NAME = DotName.createSimple(Map.class.getName());
     public static final Type MAP_TYPE = Type.create(MAP_INTERFACE_NAME, Type.Kind.CLASS);
@@ -94,6 +97,7 @@ public class OpenApiDataObjectScanner {
     public static final Type ARRAY_TYPE_OBJECT = Type.create(DotName.createSimple("[Ljava.lang.Object;"), Type.Kind.ARRAY);
 
     private static ClassInfo collectionStandin;
+    private static ClassInfo iterableStandin;
     private static ClassInfo mapStandin;
 
     /*-
@@ -103,9 +107,11 @@ public class OpenApiDataObjectScanner {
     static {
         Indexer indexer = new Indexer();
         index(indexer, "CollectionStandin.class");
+        index(indexer, "IterableStandin.class");
         index(indexer, "MapStandin.class");
         Index index = indexer.complete();
         collectionStandin = index.getClassByName(DotName.createSimple(CollectionStandin.class.getName()));
+        iterableStandin = index.getClassByName(DotName.createSimple(IterableStandin.class.getName()));
         mapStandin = index.getClassByName(DotName.createSimple(MapStandin.class.getName()));
     }
 
@@ -280,12 +286,16 @@ public class OpenApiDataObjectScanner {
 
     // Is Map, Collection, etc.
     private boolean isSpecialType(Type type) {
-        return isA(type, COLLECTION_TYPE) || isA(type, MAP_TYPE);
+        return isA(type, COLLECTION_TYPE) || isA(type, ITERABLE_TYPE) || isA(type, MAP_TYPE);
     }
 
     private ClassInfo initialType(Type type) {
         if (isA(type, COLLECTION_TYPE)) {
             return collectionStandin;
+        }
+
+        if (isA(type, ITERABLE_TYPE)) {
+            return iterableStandin;
         }
 
         if (isA(type, MAP_TYPE)) {

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeProcessor.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeProcessor.java
@@ -18,6 +18,7 @@ package io.smallrye.openapi.runtime.scanner.dataobject;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.ARRAY_TYPE_OBJECT;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.COLLECTION_TYPE;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.ENUM_TYPE;
+import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.ITERABLE_TYPE;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.MAP_TYPE;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.OBJECT_TYPE;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.STRING_TYPE;
@@ -144,6 +145,11 @@ public class TypeProcessor {
             return ARRAY_TYPE_OBJECT;
         }
 
+        // Raw Iterable
+        if (isA(type, ITERABLE_TYPE)) {
+            return ARRAY_TYPE_OBJECT;
+        }
+
         // Raw Map
         if (isA(type, MAP_TYPE)) {
             return OBJECT_TYPE;
@@ -166,7 +172,7 @@ public class TypeProcessor {
         Type typeRead = pType;
 
         // If it's a collection, we should treat it as an array.
-        if (isA(pType, COLLECTION_TYPE)) { // TODO maybe also Iterable?
+        if (isA(pType, COLLECTION_TYPE) || isA(pType, ITERABLE_TYPE)) {
             LOG.debugv("Processing Java Collection. Will treat as an array.");
             Schema arraySchema = new SchemaImpl();
             schema.type(Schema.SchemaType.ARRAY);

--- a/implementation/src/test/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScannerTestBase.java
+++ b/implementation/src/test/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScannerTestBase.java
@@ -35,6 +35,7 @@ public class OpenApiDataObjectScannerTestBase extends IndexScannerTestBase {
 
         // Stand-in stuff
         index(indexer, "io/smallrye/openapi/runtime/scanner/CollectionStandin.class");
+        index(indexer, "io/smallrye/openapi/runtime/scanner/IterableStandin.class");
         index(indexer, "io/smallrye/openapi/runtime/scanner/MapStandin.class");
 
         // Test samples

--- a/tck/src/test/java/io/smallrye/openapi/tck/extra/JaxRsTest.java
+++ b/tck/src/test/java/io/smallrye/openapi/tck/extra/JaxRsTest.java
@@ -37,7 +37,7 @@ public class JaxRsTest extends BaseTckTest<JaxRsTest.JaxRsTestArquillian> {
     public static class JaxRsTestArquillian extends AppTestBase {
         @Deployment(name = "jaxrs")
         public static WebArchive createDeployment() {
-            return ShrinkWrap.create(WebArchive.class, "airlines.war")
+            return ShrinkWrap.create(WebArchive.class, "jaxrstest.war")
                     .addPackages(true, new String[] { "io.smallrye.openapi.tck.extra.jaxrs" });
         }
 
@@ -61,5 +61,16 @@ public class JaxRsTest extends BaseTckTest<JaxRsTest.JaxRsTestArquillian> {
             ValidatableResponse vr = this.callEndpoint(type);
             vr.body("paths.'/jaxrs/widgets/{widgetId}'.put.responses.'204'.description", equalTo("No Content"));
         }
+
+        @RunAsClient
+        @Test(dataProvider = "formatProvider")
+        public void test200IterableResponse(String type) {
+            ValidatableResponse vr = this.callEndpoint(type);
+            vr.body("paths.'/jaxrs/widgets/iter'.get.responses.'200'.description", equalTo("OK"));
+            vr.body("paths.'/jaxrs/widgets/iter'.get.responses.'200'.content.'*/*'.schema.type", equalTo("array"));
+            vr.body("paths.'/jaxrs/widgets/iter'.get.responses.'200'.content.'*/*'.schema.items.'$ref'",
+                    equalTo("#/components/schemas/Widget"));
+        }
+
     }
 }

--- a/tck/src/test/java/io/smallrye/openapi/tck/extra/jaxrs/WidgetResource.java
+++ b/tck/src/test/java/io/smallrye/openapi/tck/extra/jaxrs/WidgetResource.java
@@ -47,6 +47,21 @@ public class WidgetResource {
         return rval;
     }
 
+    @GET
+    @Path("iter")
+    public Iterable<Widget> getAllWidgetsAsIterable() {
+        List<Widget> rval = new ArrayList<>();
+        Widget w1 = new Widget();
+        w1.setName("Widget One");
+        w1.setDescription("The description of widget one.");
+        Widget w2 = new Widget();
+        w2.setName("Widget Two");
+        w2.setDescription("The description of widget two.");
+        rval.add(w1);
+        rval.add(w2);
+        return rval;
+    }
+
     @POST
     public void createWidget(Widget widget) {
         // Add the widget to the DB

